daisy_workflows/linux_common/utils/guestfsprocess.py

```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/env python3
+# Copyright 2020 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Run a process in a guestfs.GuestFS instance.
+
+This library addresses a shortcoming of GuestFS.command and GuestFS.sh
+that cause data loss. When the command is successful, stderr is discarded.
+When the command fails, stdout is discarded.
+
+This library is motivated by subprocess.run, which returns an object
+containing stdout, stderr, and the return code of the process.
+"""
+
+import os
+import shlex
+import textwrap
+import typing
+
+
+def run(g: 'GuestFSInterface', command) -> 'CompletedProcess':
+  """Runs a process in a mounted GuestFS instance, ensuring that
+  standard output and standard error is always retained.
+
+  Args:
+    g: Mounted GuestFS instance.
+    command (str or List[str]): Script content that will be executed
+    by a bash interpeter on the guest.
+
+  Examples:
+    >>> run(g, 'date').stdout
+    Thu 05 Nov 2020 06:53:55 PM PST
+
+    >>> run(g, 'printf hi; exit 1')
+    {'stdout': 'hi', 'stderr': '', 'code': 1, 'cmd': 'printf hi; exit 1'}
+  """
+  tmp_dir = g.mkdtemp('gprocXXXXXX')
+  program_path = os.path.join(tmp_dir, 'program.sh')
+  stdout_path = os.path.join(tmp_dir, 'stdout.txt')
+  stderr_path = os.path.join(tmp_dir, 'stderr.txt')
+  return_code_path = os.path.join(tmp_dir, 'return_code.txt')
```

Comment:
i may be missing some unique requirements here but programs on linux can pipe their stdout, stderr etc. file descriptors to each other. it should not be necessary to create temporary files to capture these.

Reply:
This Python code runs on a worker instance, but `command` is executed in the guest. 

Typically we've been using `guestfs.sh` or `guestfs.command` for this functionality. The problem with these APIs, though, is that you get *either* stdout, or stderr, depending on the return code of the program. Specifcally:

1. When the return code is zero, return a string containing stdout.
2. When the return code is nonzero, throw an exception that captures stderr.

The goal is to have stdout, stderr, and the return code for every execution.